examples/picamera/pitrack.py

```diff
@@ -0,0 +1,182 @@
+            else:
+                payload = data[offset : offset + package_size]
+                offset += package_size
+
+            if offset == len(data):
+                fu_header = fu_header_end
+
+            packages.append(fu_header + payload)
+
+            fu_header = fu_header_middle
+        assert offset == len(data), "incorrect fragment data"
+
+        return packages
+
+    @staticmethod
+    def _packetize_stap_a(
+        data: bytes, packages_iterator: Iterator[bytes]
+    ) -> Tuple[bytes, bytes]:
+        counter = 0
+        available_size = PACKET_MAX - STAP_A_HEADER_SIZE
+
+        stap_header = NAL_TYPE_STAP_A | (data[0] & 0xE0)
+
+        payload = bytes()
+        try:
+            nalu = data  # with header
+            while len(nalu) <= available_size and counter < 9:
+                stap_header |= nalu[0] & 0x80
+
+                nri = nalu[0] & 0x60
+                if stap_header & 0x60 < nri:
+                    stap_header = stap_header & 0x9F | nri
+
+                available_size -= LENGTH_FIELD_SIZE + len(nalu)
+                counter += 1
+                payload += pack("!H", len(nalu)) + nalu
+                nalu = next(packages_iterator)
+
+            if counter == 0:
+                nalu = next(packages_iterator)
+        except StopIteration:
+            nalu = None
+
+        if counter <= 1:
+            return data, nalu
+        else:
+            return bytes([stap_header]) + payload, nalu
+
+    @classmethod
+    def _packetize(cls, packages: Iterator[bytes]) -> List[bytes]:
+        packetized_packages = []
+
+        packages_iterator = iter(packages)
+        package = next(packages_iterator, None)
+        while package is not None:
+            if len(package) > PACKET_MAX:
+                packetized_packages.extend(cls._packetize_fu_a(package))
+                package = next(packages_iterator, None)
+            else:
+                packetized, package = cls._packetize_stap_a(package, packages_iterator)
+                packetized_packages.append(packetized)
+
+        return packetized_packages
+
+    @staticmethod
+    def _split_bitstream(buf: bytes) -> Iterator[bytes]:
+        # TODO: write in a more pytonic way,
+        # translate from: https://github.com/aizvorski/h264bitstream/blob/master/h264_nal.c#L134
+        i = 0
+        while True:
+            while (buf[i] != 0 or buf[i + 1] != 0 or buf[i + 2] != 0x01) and (
+                buf[i] != 0 or buf[i + 1] != 0 or buf[i + 2] != 0 or buf[i + 3] != 0x01
+            ):
+                i += 1  # skip leading zero
+                if i + 4 >= len(buf):
+                    return
+            if buf[i] != 0 or buf[i + 1] != 0 or buf[i + 2] != 0x01:
+                i += 1
+            i += 3
+            nal_start = i
+            if (buf[nal_start] & 0x1F) != 7 and (
+                buf[nal_start] & 0x1F
+            ) != 8:  # Assuming only SPS/PPS are packaged with other NALs
```

Comment:
So this just bundles in all the remaining NALs if the NAL at index `i` is not a SPS or PPS NAL? Why is this safe to do? Is there any reading you can point me to?

This has some performance benefits as well since you skip the rest of the method, so does it make sense to bring this into the `h264.py` version?

Reply:
This might be specific to the pi camera since the picamera encoder only output a single NAL for NALs encoding images or three NALs SPS/PPS/Image per call. Other encoders may output multiple NALs at once (a full GOP or NALs for example) so there is a need to  continue splitting the bitstream after an image NAL is discovered.